### DISCUSSION TITLE
Column level lineage

### DISCUSSION THREAD

**Discussion by sou-joshi** [dbt-labs/dbt-core#4458] [Ideas]
### Describe the feature
Currently have table/view level lineage captured, can this be extended to field/column based?

### Additional context
Not specific to a database, applies to the product.

### Who will this benefit?
There could be multiple renames to a field and with multiple joins the complexity to track field level changes and source of a particular column becomes difficult. From an Ops perspective it always good to know where a field is coming from to quickly solve data issues.

**Comment by jtcohen6** [comment]
@dataexpertz-blr Thanks for opening, I'm surprised there wasn't already an issue for this :) It's something we're hearing and talking about a lot these days.

### Mechanisms

I view column-level lineage as existing in two orders of complexity:
- **Extending existing constructs.** Today, dbt developers have to duplicate a lot of resource properties (descriptions, tags, meta, tests) across models, even when model Y is just `select * from` model X. Over in #2995, and #1158 before then, we've been discussing ways that YAML anchors make a version of this possible today, and how it could be better in the future (cross-file anchors, souped-up `docs` blocks, hierarchical properties-as-config).
- **Massive new capabilities.** To capture column-level lineage for real for real, we'd need a validating SQL grammar—same as would, incidentally, for a decent linter / auto-formatter (#2356). In a world where we had this, and built it into dbt, we'd also have an AST representation of every column name, `from` relation, and SQL function. This is definitely on our minds; it's still a ways away.

### Use cases

As with any compelling feature, column-level lineage feels both immensely valuable and a bit vague. If dbt could produce an EXPLAIN-style plan, of every single SQL function performed to produce a single column, that would be very cool, and also tricky to read and reason about as a human being.

So I do find it useful to think concretely about the kinds of things we'd hope to enable here:

- **Property (tag/meta/etc) inheritance.** E.g. If a source PII/PHI column is the indirect input to a column in a downstream model, being able to mark the latter model as sensitive.
- **Saving code (+ time).** Extending what I described above, if a column has not transformed from model X to model Y—no renames, no aggregations—dbt could natively inherit its properties, such as description + tests.
    - Or, better yet, dbt would understand that the column has been tested upstream, has not changed, and so does not need to run the same tests again.
- **"dbt column advisor."** If dbt has a full picture of how a column is produced—which input columns, which transformations—it could flag when there are potentially duplicative columns across models, and help avoid the repeating of business logic.
- **Slim CI to the max:** If you've only changed one column across a few models, rather than running the changed models and _all_ their children, you'd only need to run + test downstream models that use the affected column.

I'm curious to hear what other things come to mind!

**Comment by bashyroger** [comment]
+4 for the use case you've mentioned @jtcohen6 , I'd like to add the following, IMO important use case:

- A variant of **Slim CI to the max:** being able to do an impact analysis at _development time_ when you're changing a  source / model's column name. After all, the fact that a column changes _does not necessarily mean_ that the column also **used** in any dependent models. For that, you really need column level lineage.
 
Instead of building this yourself, you could also think about integrating DBT with the only data lineage focused SAAS product I have heard of: https://getmanta.com/integrations/

**Comment by tufanrakshit** [comment]
we are planning to use DBT Cloud for our project and this table by table lineage is really a killer feature which we would like to have  as This would make data lineage and debugging much much easier

**Comment by devstein** [comment]
Hi 👋 , does anyone know how [Datafold + DBT](https://www.datafold.com/blog/datafold-dbt-the-perfect-stack-for-reliable-data-pipelines) claims to provide this?

**Comment by jaypeedevlin** [comment]
Datafold use their own lineage capabilities to do this — while they do read in your dbt project, the column level lineage is part of their platform (it's an awesome feature though!)

**Comment by MarkMacArdle** [comment]
Monzo (a UK startup bank) have [written about](https://medium.com/data-monzo/mapping-our-data-journey-with-column-lineage-56209c00606d) how they created column lineage using BigQuery audit logs and parsing the compiled dbt queries with [ZetaSQL](https://github.com/google/zetasql). Hadn't heard about ZetaSQL before but it's an open source project from Google for analysing SQL.

That solution is BigQuery specific but still interesting to see an implemented approach to this problem.

**Reply by after-ephemera** [reply]
[Here](https://cloud.google.com/architecture/building-a-bigquery-data-lineage-solution#extracting_lineage) is GCP's post describing a similar approach.

**Comment by christopherridley** [comment]
Is there a huge benefit here where you could define your docs/descriptions once at the source table level and then have that trickle down to all columns that reference the original source column? Still wrestling how to populate descriptions for all columns effectively and efficiently.

**Reply by KayakinCoder** [reply]
This, in my opinion, is a huge unmet need. Documenting data has become a huge time suck, and the lack of description propagation is a reason why.

**Reply by tripleaceme** [reply]
Hello @christopherridley and @KayakinCoder,

I'm working on something to solve this, as our dbt project is also growing, and copying descriptions around is becoming too much of a headache.

I'll share more updates in the coming days.

Thank you

**Reply by VDFaller** [reply]
[dbt-osmosis](https://github.com/z3z1ma/dbt-osmosis) does that.  I also included it in my [rust based fusion thing](https://github.com/VDFaller/dbt-lint-yaml) that is eventually meant to be sort of like a ruff for dbt projects.   I'm only the developer of the latter project.  But I can't live without the former.

**Reply by tripleaceme** [reply]
I'm not working on column lineage problem but the description trickling down

**Reply by VDFaller** [reply]
Yeah that's what osmosis does.

**Comment by jtalmi** [comment]
one new application i thought of: dbt build could only block downstream nodes that are directly affected by a test failure. let's say a random column column1 fails in model A, and model A -> model B. B uses column2 from A but not column1. if a test on column1 fails, let's say a not null test, it shouldn't block model B since model B doesn't use column1.

**Comment by bashyroger** [comment]
It looks like the following company is working on this functionality: https://www.hivedive.io/
Did anyone testdrive this already?

**Comment by stephanclaus** [comment]
I think there are countless vendors out there specializing on a certain database vendor (i.e. for Snowflake column level lineage is easier as they provide great meta data around accessed objects on column level already).

Another very concrete use case I wanted to add: Optimize your tracking plan (and thus, saving compute)

We are using the open source version of Snowplow and are lacking a unified UI for frontend + backend + analytics to align for. That lead to situations where very similar events where added but nothing was deprecated as "it could break things downstream". Being able to analyse which events from the raw source tables are used in downstream models would make that way more convenient.

**Comment by mertbakir** [comment]
I was expecting dbt to generate (inherit) column descriptions for downstream tables if it's already defined on source table. Then I faced this discussion and realised I also need column level lineage because it takes so much time to understand the affect of changing a column.

**Comment by vergenzt** [comment]
I address some of these limitations (specifically the maintenance ones, i.e. the ability to "forward" unchanged column descriptions and tests from upstream to downstream models -- this doesn't expose that lineage anywhere) with some dbt macros on top of the scripts I defined for https://github.com/dbt-labs/dbt-core/discussions/5093#discussioncomment-3159441. It's a bit convoluted so not going to go into full detail right now unless folks ask.

Summary: After I had a solution for #5093 via SQL comments & a regex extraction script, I started using a modified version of [`dbt_utils.star`](https://github.com/dbt-labs/dbt-utils/blob/main/macros/sql/star.sql) (I called my macro `ref_columns_from`) to pull from [`graph.nodes`](https://docs.getdbt.com/reference/dbt-jinja-functions/graph) instead of the database, so I can get _dbt_'s metadata about that model and its columns instead of just what the database knows about it from [`get_columns_in_relation`](https://docs.getdbt.com/reference/dbt-jinja-functions/adapter#get_columns_in_relation). Then, since a single `dbt compile` pass can only propagate columns one level, I iterate through "dbt compile" → extract column definitions from compiled SQL → "dbt compile" → ..., until the generated "schema.yml" file (column doc & test definition) stops changing.

**Reply by z3z1ma** [reply]
https://github.com/z3z1ma/dbt-osmosis

This tool does it already with a very easy CLI @vergenzt

**Comment by YohanGrember** [comment]
I'm currently ramping up on an existing DBT stack, and I just wanted to stress how column-level lineage would help me do so more efficiently.

Without it, figuring out the upstream flow of a transformed table field takes me forever, and I'd dream to be able to see it in one click! This would be a game-changer for my productivity and I hope we'll move forward on this topic 🙏

**Comment by ijoseph** [comment]
[Amundsen](https://medium.com/amundsen-io/amundsen-monthly-update-april-2021-ab617c20a6d6) ([second link](https://atlan.com/amundsen-data-lineage-set-up/)) is open-source, perhaps someone can try it out in a real-world environment and report back (so I don't have to be the guinea pig) :) 

> We have added native lineage support in Amundsen so you can ingest lineage metadata (both at table and **column** level) straight into the graph backend.

**Reply by after-ephemera** [reply]
I've played with Amundsen and it's great for bringing lineage in, but I think this issue was specifically focused on capturing column-level lineage within dbt, which is a bit of a different beast.

**Comment by d-axel-b** [comment]
@jtcohen6, any news about doing it in DBT directly?

**Comment by jtcohen6** [comment]
_Hey all, it's been a while since I provided an update on this! And it continues to be something that we talk about a lot with users and ecosystem partners. I'm going to share below a lightly edited version of an answer that I gave to this question recently, when chatting about it with some other folks building tools on top of dbt. If you read both our [May roadmap](https://github.com/dbt-labs/dbt-core/blob/main/docs/roadmap/2022-05-dbt-a-core-story.md#:~:text=Building%20a%20SQL,template%20model%20SQL.) and [August roadmap](https://github.com/dbt-labs/dbt-core/blob/main/docs/roadmap/2022-08-back-for-more.md#:~:text=1Now%2C%20I,each%20one%20shines.) for dbt Core, then the evolution I'm describing below will sound familiar._

### Background

There was a time, not that long ago, when it felt like a (Jinja-)SQL grammar was of the the most important things we could be building into `dbt-core`. Also, when asked "What is dbt?," one of my go-to answers used to be, "It's a programming framework for SQL."

The two things that have changed since then are:
1. Our vision about what dbt is, exactly: More Than Just SQL :tm:. There's increasing appetite for other modeling languages in dbt — we just added Python, but we've also heard interest in opinionated frameworks that compile down to SQL, such as [PRQL](https://github.com/dbt-labs/dbt-core/discussions/5796) or [Ibis](https://github.com/dbt-labs/dbt-core/discussions/5738).
2. Developments in the ecosystem. In the last few years, there has been a lot of impressive OSS work (not by us) on SQL parsers/analyzers/transpilers: [sqlparser-rs](https://github.com/sqlparser-rs/sqlparser-rs), [sqlglot](https://github.com/tobymao/sqlglot), ...

I want users to be able to leverage the work that others have been doing on SQL parsers, while we haven't been able to prioritize it ourselves — and to make the most of modeling languages that offer up column-aware ASTs "for free."

### Building it all ourselves -> enabling an ecosystem

So, when I think about the highest-leverage things we can do here, I think our attitude has shifted somewhat, from:

"We must ourselves build a validating grammar for Jinja-SQL"

to

"We can enable this future by supporting more modeling languages in dbt — including user-friendly languages that compile down to SQL — and letting users get the most of them."

For what Jinja-SQL gets you in terms of dynamic SQL templating (`run_query` -> use results to template out model SQL), there's a necessary trade-off in terms of tracking column-level lineage in advance of a model running. @vergenzt gets at this by mentioning the switch from `dbt_utils.star` (dynamic compilation at runtime, requires database state as an input) to custom macros leveraging `graph` (static information, knowable at parse time). It won't surprise anyone to hear that Python can be even wilder. By contrast, other modeling languages are more natural fits for statically analyzing column relationships.

That's the dream, right? As far as doing smart column-aware things, the earlier dbt can know, the more we can do at runtime: select only models to run that are downstream of columns with tag pii; cascade that tag down to all child columns at the end of parsing; skip running models in CI that depend on changed models but only unchanged columns; etc.

Otherwise, it's just richer metadata on already-compiled queries / already-run models. Many ecosystem partners have built tooling for this — which is sweet! — by passing `compiled_code` from each model in `manifest.json` through a SQL grammar, and offering compelling ways to visualize / explore the result.

I don't know exactly how all these pieces will fit together. If your modeling language / transpiler is column-aware, how should we be exposing those capabilities within dbt? We're very interested in exploring this more over the next several months, as we also work on refactoring the tight coupling of Jinja-SQL within dbt, and creating a more-pluggable interface for modeling languages.

### Back to basics

To be clear, Jinja-SQL isn't going away anytime soon. The goal would be to support it _in addition to_ other modeling languages, and to share _really clear opinions_ about their relative strengths and weaknesses. (Check out [our Coalesce talk](https://www.youtube.com/watch?v=rVprdyxcGUo) for our attempt at a "fun" version of this, presenting SQL "versus" Python.)

And we will continue to invest in making the Jinja-SQL development experience better and better, since it's what 99% of dbt models are written in today. If someone does build a really good version of this — a validating Jinja-SQL grammar that can analyze model code at parse time, or even just a SQL grammar that can analyze code at compile time — _and_ if we want all users to benefit from it, for all tools in the ecosystem to have access to it, for it to become part of the dbt standard — then pieces of it may eventually want to live inside the core framework of dbt.

I'm going to keep writing in public about our vision for a more-pluggable `dbt-core`. We're spending the next few months laser-focused on laying the technical foundations to make it possible, to provide worthy interfaces for integration. In the meantime, if this is something you're interested in working on: have at it! I'm excited to see what you come up with :)

**Reply by calleo** [reply]
I like this vision and want to read about all the modelling languages you mention.

It also sounds a lot familiar to frontend development using JavaScript.

There are different web browsers (databases), supporting different versions of JavaScript (SQL), sometimes with their own custom additions.

In addition to that we have languages that transpile into JavaScript (i.e. Typescript), that are supposed to bridge this gap.

Historically there has been other tools trying to turn this into an acceptable developer experience (write once, run everywhere). For example webpack, Babel. 

My own personal opinion from working with this, struggling with it, and seeing others struggle is that the ecosystem has become so damn complicated that very few understands what's going on.

What I want to say is, that you can probably learn alot by looking at the JavaScript/Typescript community and seeing where they are heading. And not repeat the same mistakes 😃

**Comment by ammarchalifah-bolt** [comment]
It's been a couple of months since last update on this topic. Is there any update on dbt-core development regarding column lineage feature?

**Comment by jmbrooks** [comment]
I'd like to add the very high value this can have for folks using dbt to build feature stores in their data warehouse for data science/ML use cases!

Feature built and defined in an offline feature store (& registry) function best when there is column-level lineage on those features, because any upstream transformations or business logic changes require new "versions" of those features. Without column-level lineage, it can be more difficult to maintain a feature store built in dbt, as tracking changes of particular columns is not precise. Excited to potentially see this in dbt someday!

**Comment by JoePollo** [comment]
Enterprise compliance for columnar lineage is causing heartache for my corp as we adopt DBT; this is a well desired feature that would provide exponential value to us. Native DBT functionality in the capacity of [DBTerd ](https://github.com/datnguye/dbterd) would be fantastic. We are working on creating a templated repo that would be used by hundreds of engineers across the corporation, all of which would need to support columnar lineage.

**Comment by nicoteiza** [comment]
Hi! I'd love to read if there has been any update in this front. As I understand [sqlmesh](https://github.com/TobikoData/sqlmesh) can do column lineage. Does anyone know how they achieved it? Is it something that could be translated somehow to dbt?

**Reply by indy-jonesy** [reply]
The creator of  sqlmesh, also created [sqlglot](https://github.com/tobymao/sqlglot).
You can find an independent dev team were able to build a proprietary dbt lineage solution [here](https://www.reddit.com/r/dataengineering/comments/11llsqn/building_column_level_lineage_for_dbt/) using [sqlglot as a backbone](https://www.reddit.com/r/dataengineering/comments/11llsqn/comment/jbfa2sj/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button). It would be a matter of leveraging sqlglot as a compiler and incorporating it into dbt core.

**Reply by ismailsimsek** [reply]
@indy-jonesy @nicoteiza it seems like dbt proprietary feature is also based on sqlglot  [link1]( https://www.reddit.com/r/dataengineering/comments/1aqcvov/dbt_column_level_lineage_only_for_dbt_cloud/) [link2](https://twitter.com/captaintobs/status/1757601463852023876)

**Comment by VDFaller** [comment]
Would [ColumnInfo](https://github.com/dbt-labs/dbt-core/blob/23b16ad6d2f90f4d79ceb4a88d0efa412bf90abf/core/dbt/contracts/graph/nodes.py#L216) be the right place to try to put this in the manifest? Or am I barking up the wrong tree?  Would it be helpful if we could get something like 

```json
{
    "nodes": {
        "model.my_dbt.child_model": {
            "database": "dev_vince",
            "schema": "my_schema",
            "name": "child_model",
            "resource_type": "model",
            "package_name": "mydbt",
            "columns": {
                "id": {
                    "name": "id",
                    "description": "my_pk",
                    "depends_on": []                    
                },
                "source_column": {
                    "name": "source_column",
                    "description": "some column passed through from a source table",
                    "depends_on": [
                        {"node": "model.my_dbt.parent_source", "column": "source_column"}
                    ]
                },
                "model_column": {
                    "name": "model_column",
                    "description": "some column passed through from a model",
                    "depends_on": [
                        {"node": "model.my_dbt.parent_model", "column": "model_column"}
                    ]
                }, 
                "surrogate_key": {
                    "name": "surrogate_key",
                    "description": "this is a derived surrogate key from source_column/model_column",
                    "depends_on": [
                        {"node": "source.my_dbt.source_schema.parent_source", "column": "source_column"}, 
                        {"node": "model.my_dbt.parent_model", "column": "model_column"}
                    ]
                }
            }
        }
    }
}
```

**Comment by shishircc** [comment]
Is this on roadmap ? SQLMesh is able to use dbt project as source and generate column level lineage. Since that is open source, could that be used to bring the column level lineage to DBT ?

**Comment by shishircc** [comment]
The underlying library, sqlglot makes this very easy. 
E.g. See the simple 3 lines of python code creating column level lineage graph here https://twitter.com/Captaintobs/status/1619185992111624197

**Reply by robomill** [reply]
It's definitely more complicated than this depending on the complexity of the sql. But this is essentially how datahub generates lineage:

https://github.com/datahub-project/datahub/blob/8415fc214b1b9d0f99eafbffc01f3e46043f54dc/metadata-ingestion/src/datahub/utilities/sqlglot_lineage.py#L178

**Reply by ismailsimsek** [reply]
@robomill it seems like dbt proprietary feature is also based on sqlglot [link1](https://www.reddit.com/r/dataengineering/comments/1aqcvov/dbt_column_level_lineage_only_for_dbt_cloud/) [link2](https://twitter.com/captaintobs/status/1757601463852023876)

**Reply by robomill** [reply]
@ismailsimsek It's kinda weird they didn't implement it themselves for enterprise. 

They already have all the ingredients to write it themselves built into DBT. They already have the sqlparse library that generates an AST that they could use to generate the inputs and outputs of a model. They already have code that can introspect the schema of tables as well. It's easy enough with a good engineer to reverse engineer what SQLglot have done.

Seems kinda lazy on their part and less of a reason to exclude from dbt-core if they are using a third party open source library

**Comment by shishircc** [comment]
After the great announcements at coalesce, is this on roadmap ?

**Reply by matfior-finn** [reply]
While we wait, I suggest checking https://www.turntable.so/, it's quite nifty and really fast.

**Reply by shishircc** [reply]
Turntable is great. My main concern is that it requires developer level access to get the lineage in VS Code. All people who benefit from lineage may not have developer level access.

**Reply by ianrtracey** [reply]
@shishircc Ian from Turntable here. What do you mean by developer level access? We're currently building a cloud version if you'd like to learn more

**Comment by matteoannotell** [comment]
Hello friends, is there any updates on this one?

**Reply by matfior-finn** [reply]
I would like to know as well. I tried asking during coalesce in London but I got some evasive answers.

**Comment by AlexThomas90210** [comment]
Is progress still being made on this?

**Reply by rrivera-ut** [reply]
I'm going to take a stab and say probably not

**Reply by matfior-finn** [reply]
would be cool to get some sort of official answer about it.

**Reply by ViniciusRaphael** [reply]
That's true. Some official answer about status would be great.

**Comment by borjavb** [comment]
If anyone is interested into building something for BigQuery, this parser will give you full column lineage for your queries, it covers pretty much the whole BigQuery syntax! https://github.com/borjavb/bq-lineage-tool

**Comment by mertbakir** [comment]
It looks like dbt decided to make column level lineage for the cloud only. That's sad.

**Reply by matfior-finn** [reply]
Enterprise only, that's sad

**Reply by ViniciusRaphael** [reply]
Column level lineage on dbt cloud? Just accessed there and couldn't find anything like that. Are you guys sure?

**Reply by matfior-finn** [reply]
see here: https://www.getdbt.com/blog/proactively-improve-your-dbt-projects-with-new-dbt-explorer-features

> Column-level lineage is available in public beta to Enterprise plan customers only.

**Reply by ViniciusRaphael** [reply]
Thanks @matfior-finn

**Reply by indy-jonesy** [reply]
Sad indeed.

**Reply by odikia** [reply]
Are they trying to push developers toward sqlmesh? Cause locking the critical CLL feature behind a MASSIVE paywall is how you do it...

**Reply by ianrtracey** [reply]
For anyone looking for CLL, we've build a free extension for VS Code

https://marketplace.visualstudio.com/items?itemName=turntable.turntable-for-dbt-core

**Reply by odikia** [reply]
Hey @ianrtracey , I do enjoy turntable, but I'm wondering if enterprise security would be happy with data transfer to outside servers for information parsing? Some of the information in most stacks regarding database objects is proprietary, and if they're not happy with sending data to OpenAI via API (leading to current trend in local LLM deployment), I don't think they'd be okay with sending data to Turntable's servers for LLM interaction. It's not entirely clear how security teams at an enterprise can evaluate the transfer of data at present with Turntable given the tool's present documentation (that I am aware of, at least).

**Reply by ianrtracey** [reply]
@odikia everything in Turntable runs locally on your machine (besides LLM auto-documentation generation). This includes column-level lineage, the docs GUI, query preview, and language server features (unlike dbt-power user which I believe does send metadata to servers for column-level lineage).

Unless you click "auto generate docs", nothing ever leaves your device. You can also disable this feature (as well as all server communication besides auth) in our settings panel. We built this setting recently to support some folks at Anduril (which have intense info sec requirements).

We're also more than happy to provide source code for teams that want that extra peace of mind as well (and we're also considering open sourcing our extension this year).

<img width="864" alt="Screenshot 2025-01-11 at 2 55 29 PM" src="https://github.com/user-attachments/assets/2ba0a17f-e40a-424f-93c9-a07b819ed019" />

**Comment by rrivera-ut** [comment]
Is anyone surprised?

**Reply by calleo** [reply]
I don't understand the disappointment.

Most of us are working for companies that needs to earn money to keep on living. DBT is no exception. Why on earth should they be giving away the product for free?

Of course we can go build something new and free, but it will take a long time, and eventuelly we will also have to pay our bills.

**Reply by KayakinCoder** [reply]
@calleo I'm a bit confused. The Team plan is $100/month _per developer_, and as of right now it doesn't include column level lineage. You consider that pricing to be free?

Let's all just hope that the folks at dbt realize that making column-level lineage available to SMBs on the Team plan would add a lot of value/use cases that could justify teams purchasing plans.

**Comment by rrivera-ut** [comment]
DBT is just too expensive - it's competitors who are building better
features are cheaper - see Paradime.io

**Comment by jyotidhiman0610** [comment]
Is there any plan to release this for dbt core?

**Reply by 1cadumagalhaes** [reply]
Pretty sure there isn't

**Comment by ismailsimsek** [comment]
Would it be accepted if community adds this feature to dbt core? im happy to help.

**Reply by 1cadumagalhaes** [reply]
I was thinking about that this week, maybe creating a dbt package that adds column level lineage would work.
I'm not sure if this could be added to the core itself.
It should be possible since both dbt (on cloud) and sqlmesh have column level lineage

**Reply by ismailsimsek** [reply]
+1
also it looks like both (dbt on cloud and sqlmesh) are using opensource sqlglot for column level lineage [link1](https://www.reddit.com/r/dataengineering/comments/1aqcvov/dbt_column_level_lineage_only_for_dbt_cloud/) [link2](https://twitter.com/captaintobs/status/1757601463852023876)

side note: AFAIU using `sqlglot`, its also possible to get rid of the `{{ref('model_a')}}` syntax. and use pure sql(table name) instead. that's how sqlmesh find dependencies.

**Reply by z3z1ma** [reply]
A few thoughts. At this point just use sqlmesh. Really. Why jump through these hoops. Dbt implemented CLL (and the backing UI) as a paid feature which isn't inherently wrong. They are a business at the end of the day. It would not benefit them to pursue this in the oss product after releasing it as a differentiator in the paid product. Either it's worst and people are unhappy and they must maintain it or they devalue their cloud products value proposition. So probably just make another pypi package if you want to attempt it. 

And doing away with refs is shortsighted only because people override the ref macro constantly and it's used in so many places. Jinja and dbt are quite married at this point if you dig under the hood. Furthermore people depend on dbt behaving consistently and not discover new "refs" because of some new parsing.

Sqlglot is not just 'some oss package' with a one way relationship to sqlmesh but is in fact built and maintained by the creators of sqlmesh. So their usage of it is in the blood of the software and not bolted on.

**Comment by wenwu35** [comment]
We just released an open-source [utility](https://github.com/canva-public/dbt-column-lineage-extractor) for dbt column-level lineage.

This was primarily developed to provide a programmatic interface to dbt column level lineage, enabling further development and use, such as creating automated tools for tagging sensitive column data. But you can also visualized the output json using tools such as [jsoncrack.com](https://jsoncrack.com/editor)

Existing column-level lineage tools, like Atlan, dbt Cloud, SQLMesh, and Turntable, lack this programmatic capability and also face challenges such as subscription fees, indexing delays, complexity, or concerns about transmitting organizational code/data to vendor servers, which hinder their wider adoption.

Hope it helps.

**Reply by ismailsimsek** [reply]
@wenwu35 Would it be possible to add this feature to [opendbt](https://github.com/memiiso/opendbt). opendbt already adds some [new features](https://github.com/memiiso/opendbt/blob/main/docs/EXAMPLES.md) to dbt-core and this will be great addition. opendbt is `dbt-core`+additional features which are not accepted in dbt-core. i believe this will be good addition. whoever is interested contributions are welcome.

i am in progress of adding [sql formatting to it](https://github.com/memiiso/opendbt/pull/18)

**Reply by 1cadumagalhaes** [reply]
Hey, this looks like an interesting project. Some of the features that are there and work with airflow are in the astronomer cosmos project, maybe you could take a look

**Comment by ismailsimsek** [comment]
This feature is implemented by [opendbt](https://github.com/memiiso/opendbt) now see [documentation and the demo](https://memiiso.github.io/opendbt/)

along the [column dependencies](https://memiiso.github.io/opendbt/opendbtdocs/) you can also see the source transformations that produces the column.

and its more polished and user friendly UI, to be used by business users.

please feel free to give it a test, use it, or add improvement. opendbt welcomes any contribution.

**Comment by labnxt** [comment]
Sounds great - I will take a look when I get the chance!!

🙂

**Comment by sqlparser** [comment]
Column-level lineage in dbt has been a pain point for our team too. A few approaches we've seen work:

1. **sqlglot + custom scripting**: Works for basic cases but struggles with `SELECT *` resolution without schema context, and Oracle/SQL Server proprietary syntax.

2. **Gudu SQL Omni** ([VS Code extension](https://marketplace.visualstudio.com/items?itemName=gudusoftware.gudu-sql-omni)): If you work in VS Code, this extension provides column-level lineage directly in the editor — it auto-detects dbt projects and resolves `ref()`. Runs 100% offline, 34 dialects, free trial.

3. **OpenLineage + Marquez**: Good standard for runtime lineage capture, but doesn't do static SQL analysis.

The hard part is always the SQL parsing — different warehouses have different syntax extensions, and `SELECT *` needs schema resolution to produce accurate lineage. Happy to share specifics on how we set this up.

(Disclosure: I work at Gudu Software, which builds SQLFlow and SQL Omni.)

**Comment by Tsumu-x** [comment]
Hi, I've been following this thread for a while. Since column-level lineage                            
  in dbt-core still seems unlikely anytime soon, I ended up building my own                              
  tool to scratch this itch.                                                                             
                                                                                                         
  It's a Rust CLI called lineage-lite — does static analysis on SQL, dbt                                 
  compiled models, and Python ETL to get column-level lineage without needing
  dbt Cloud. You can trace a column back to its source and see the                                       
  transformation chain, run impact analysis before schema changes, that kind                             
  of thing.                                                                                              
                                                                                                         
  It also handles the case where your pipeline isn't pure dbt — like legacy                              
  Python jobs or ad-hoc SQL reports that dbt's lineage graph doesn't cover.
                                                                                                         
  https://github.com/Tsumu-x/lineage-lite                     
                                                                                                         
  Still early stage, but it works on our production project. Happy to hear                               
  feedback or ideas.
